crates/bitcoind_rpc/src/lib.rs

```diff
@@ -0,0 +1,321 @@
+//! mempool.
+//!
+//! # [`Iter`]
+//!
+//! [`Emitter::into_iterator<B>`] transforms the emitter into an iterator that returns
+//! [`Emission<B>`] items. The `B` generic can either be [`EmittedBlock`] or [`EmittedHeader`] and
+//! determines whether we are iterating blocks or headers.
+//!
+//! The iterator initially returns blocks/headers in increasing height order. After the chain tip is
+//! reached, the next update is the mempool. After the mempool update is released, the first
+//! succeeding call to [`Iterator::next`] will return [`None`]. Subsequent calls will resume
+//! returning [`Some`] once more blocks are found.
+//!
+//! ```rust,no_run
+//! use bdk_bitcoind_rpc::{Emission, EmittedBlock, Emitter};
+//! # let client: bdk_bitcoind_rpc::bitcoincore_rpc::Client = todo!();
+//!
+//! for r in Emitter::new(&client, 709_632).into_iterator::<EmittedBlock>() {
+//!     let emission = r.expect("todo: handle error");
+//!     match emission {
+//!         Emission::Block(b) => println!("block {}: {}", b.height, b.block.block_hash()),
+//!         Emission::Mempool(m) => println!("mempool: {} txs", m.len()),
+//!     }
+//! }
+//! ```
+#![warn(missing_docs)]
+
+use std::{collections::BTreeMap, marker::PhantomData};
+
+use bitcoin::{block::Header, Block, BlockHash, Transaction};
+pub use bitcoincore_rpc;
+use bitcoincore_rpc::bitcoincore_rpc_json;
+
+/// Represents a transaction that exists in the mempool.
+pub struct MempoolTx {
+    /// The transaction.
+    pub tx: Transaction,
+    /// Time when transaction first entered the mempool (in epoch seconds).
+    pub time: u64,
+}
+
+/// A block obtained from `bitcoind`.
+pub struct EmittedBlock {
+    /// The actual block.
+    pub block: Block,
+    /// The height of the block.
+    pub height: u32,
+}
+
+/// A block header obtained from `bitcoind`.
+pub struct EmittedHeader {
+    /// The actual block header.
+    pub header: Header,
+    /// The height of the block header.
+    pub height: u32,
+}
+
+/// A structure that emits data sourced from [`bitcoincore_rpc::Client`].
+///
+/// Refer to [module-level documentation] for more.
+///
+/// [module-level documentation]: crate
+pub struct Emitter<'c, C> {
+    client: &'c C,
+    start_height: u32,
+
+    emitted: BTreeMap<u32, BlockHash>,
+    last: Option<bitcoincore_rpc_json::GetBlockResult>,
+}
+
+impl<'c, C: bitcoincore_rpc::RpcApi> Emitter<'c, C> {
+    /// Constructs a new [`Emitter`] with the provided [`bitcoincore_rpc::Client`].
+    ///
+    /// `start_height` is the block height to start emitting blocks from.
+    pub fn new(client: &'c C, start_height: u32) -> Self {
+        Self {
+            client,
+            start_height,
+            emitted: BTreeMap::new(),
+            last: None,
+        }
+    }
+
+    /// Emit all mempool transactions.
+    pub fn mempool(&self) -> impl Iterator<Item = Result<MempoolTx, bitcoincore_rpc::Error>> + '_ {
+        let res = self.client.get_raw_mempool();
```

Comment:
side question: there a way of custom dumping the mempool so you only get transactions that entered after a certain time?

Reply:
I don't see anything in the RPC options for `get_raw_mempool` to filter what it returns. Probably the only way to make this more efficient is to plug into a `zmq` topic like `rawtx` and get block and mempool tx updates that way. But figuring out if/how to use `zmq` should be a post 1.0 thing to do.

https://bitcoincore.org/en/doc/25.0.0/rpc/blockchain/getrawmempool/

Reply:
`get_raw_mempool` only gets txids. We can use the `get_mempool_entry` on an individual txid to get the time that the tx entered the mempool.